### PR TITLE
(Archiving) Use tag when fetching spring files

### LINKED ISSUE #2729: (Archiving) Use archive location tag to filter files to archive
## Description

With the merging of #2345, very inefficient CLI functionality was added to cg. The main cause of the slow running times is that we had no option to not fetch all spring files for PDC customers (i.e. all 10^5 + spring files) and then filter the files based on the archive location by doing calls to StatusDB. #2717 gave support for a much better option since Spring files are from now on tagged with the archive location. We should make use of this to speed up the CLI commands.

## Suggested solution

Use the archive location tag when fetching Spring files so that we need not fetch and handle all Spring files belonging to PDC customers.

## This can be closed when

The archiving CLI commands are running much quicker.

### PR DESCRIPTION
**Pull request by islean**
## Description

Closes #2729. The intent is to speed up the archive CLI commands by filtering the spring files we fetch with the newly added archive location tag.

### Changed

- Files to archive are now filtered already in the SQL query, using the archive location tag.

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b archiving-use-location-tag -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

**Comment by islean** [comment]
Archiving three files using the tag reduced the run time from 20 minutes to 30 seconds:

```
θ63° 1d [isak.ohlsson@hasta:~] [S_base] 3m50s $ cg archive archive-spring-files
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/urllib3/connectionpool.py:1095: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cg-df-cat.scilifelab.se'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Sending request with headers: 
{'Content-Type': 'application/json', 'accept': 'application/json', 'Authorization': 'Bearer ****'} 
and body: 
{'osType': 'Unix/MacOS', 'createFolder': True, 'settings': [], 'pathInfo': [{'destination': 'archive@SciLifeLab:/stage/isaks_test_bundle', 'source': 'gpfs@Hasta:/proj/stage/housekeeper-bundles/isaks_test_bundle/2023-11-30/isaks_test_spring_file.spring'}], 'metadataList': []}
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/urllib3/connectionpool.py:1095: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cg-df-cat.scilifelab.se'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Files submitted to karolinska_bucket with archival task id 132.
Adding an archive to file 5002316 with archiving task id 132
θ62° 1d [isak.ohlsson@hasta:~] [S_base] 28s $ 
```

**Comment by islean** [comment]
Updating job statuses ran for 27 seconds as well:

```
θ60° 1d [isak.ohlsson@hasta:~] [S_base] 24s 2 $ cg archive update-job-statuses
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/urllib3/connectionpool.py:1095: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cg-df-cat.scilifelab.se'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Fetching status for job with id 132 from karolinska_bucket
Sending request with headers: 
{'Content-Type': 'application/json', 'accept': 'application/json', 'Authorization': 'Bearer ****'} 
and body: 
{'id': 132}
/home/proj/stage/bin/miniconda3/envs/S_cg/lib/python3.11/site-packages/urllib3/connectionpool.py:1095: InsecureRequestWarning: Unverified HTTPS request is being made to host 'cg-df-cat.scilifelab.se'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#tls-warnings
  warnings.warn(
Miria returned status Completed for job 132
Job with id 132 has finished, updating Archive entries.
θ60° 1d [isak.ohlsson@hasta:~] [S_base] 27s $ 
```

File: cg/meta/archive/archive.py [hunk 1]
```diff
@@ -81,17 +64,19 @@ def archive_spring_files_and_add_archives_to_housekeeper(
...
         for archive_location in ArchiveLocations:
-            files_and_samples_for_location: list[FileAndSample] = filter_files_on_archive_location(
-                files_and_samples=files_and_samples, archive_location=archive_location
+            if spring_file_count_limit and nr_of_files_handled > spring_file_count_limit:
+                LOG.info(f"{spring_file_count_limit} files sent to archive - exiting.")
+            files_to_archive: list[File] = self.housekeeper_api.get_non_archived_spring_files(
+                tags=[archive_location],
+                limit=spring_file_count_limit - nr_of_files_handled
+                if spring_file_count_limit
+                else None,
```
Comment (ChrOertlin) [inline cg/meta/archive/archive.py:75]: Not sure what to do here but it indeed reads quite difficult to me.

**Review by ChrOertlin — APPROVED** [review]
💯

File: cg/apps/housekeeper/hk.py [hunk 2]
```diff
@@ -418,13 +423,19 @@ def is_fastq_or_spring_in_all_bundles(self, bundle_names: list[str]) -> bool:
...
+            bundle_name=bundle_name, tags=tags or []
+        )
 
-    def get_archived_files(self, bundle_name: str, tags: list | None = None) -> list[File]:
+    def get_archived_files_for_bundle(
+        self, bundle_name: str, tags: list | None = None
+    ) -> list[File]:
         """Returns all archived_files from a given bundle, tagged with the given tags"""
-        return self._store.get_archived_files(bundle_name=bundle_name, tags=tags or [])
+        return self._store.get_archived_files_for_bundle(bundle_name=bundle_name, tags=tags or [])
```
[the remark this answers is not in the document]
Comment (islean) [reply cg/apps/housekeeper/hk.py:438]: As far as I know, all calls towards HK should go via the HousekeeperAPI in hk.py. Am I mistaken in this being a pattern we use?

**Review by Vince-janv — APPROVED** [review]
Nice solution 👍

File: cg/meta/archive/archive.py [hunk 3]
```diff
@@ -81,17 +64,19 @@ def archive_spring_files_and_add_archives_to_housekeeper(
...
-
+        nr_of_files_handled: int = 0
         for archive_location in ArchiveLocations:
-            files_and_samples_for_location: list[FileAndSample] = filter_files_on_archive_location(
-                files_and_samples=files_and_samples, archive_location=archive_location
+            if spring_file_count_limit and nr_of_files_handled > spring_file_count_limit:
+                LOG.info(f"{spring_file_count_limit} files sent to archive - exiting.")
+            files_to_archive: list[File] = self.housekeeper_api.get_non_archived_spring_files(
+                tags=[archive_location],
+                limit=spring_file_count_limit - nr_of_files_handled
```
[the remark this answers is not in the document]
Comment (islean) [reply cg/meta/archive/archive.py]: Removed nr_of_files_handled

File: cg/meta/archive/archive.py [hunk 4]
```diff
@@ -81,17 +64,19 @@ def archive_spring_files_and_add_archives_to_housekeeper(
...
-        files_and_samples: list[FileAndSample] = self.add_samples_to_files(files_to_archive)[
-            :spring_file_count_limit
-        ]
-
+        nr_of_files_handled: int = 0
         for archive_location in ArchiveLocations:
-            files_and_samples_for_location: list[FileAndSample] = filter_files_on_archive_location(
-                files_and_samples=files_and_samples, archive_location=archive_location
+            if spring_file_count_limit and nr_of_files_handled > spring_file_count_limit:
+                LOG.info(f"{spring_file_count_limit} files sent to archive - exiting.")
```
[the remark this answers is not in the document]
Comment (islean) [reply cg/meta/archive/archive.py]: Removed nr_of_files_handled

File: cg/meta/archive/archive.py (hunk 1 above)
Comment (islean) [reply cg/meta/archive/archive.py]: Went with removing the nr_of_files_handled variable altogether

**Comment by islean** [comment]
Deployed to production:

```
Log deploy... done.
cg, version 53.4.3
```

**Pull request merged**